service/storage_service.cc

```diff
@@ -3858,8 +3858,11 @@ future<> storage_service::on_change(inet_address endpoint, application_state sta
             slogger.debug("Ignoring state change for dead or unknown endpoint: {}", endpoint);
             co_return;
         }
-        if (get_token_metadata().is_normal_token_owner(endpoint)) {
-            slogger.debug("endpoint={} on_change:     updating system.peers table", endpoint);
+        const auto host_id = _gossiper.get_host_id(endpoint);
+        const auto& tm = *get_token_metadata().get_new();
+        const auto ep = tm.get_endpoint_for_host_id_if_known(host_id);
+        if (ep && *ep == endpoint && tm.is_normal_token_owner(host_id)) {
```

Comment:
So what was happening before, in the scenario of IP change, if `on_change` was called with the old IP? Do I understand correctly, it would not pass the `is_normal_token_owner(endpoint)` check?